### ISSUE TITLE
Reduce code duplication

### ISSUE THREAD

**Issue by ThibSrb** [ThibSrb/iter-tree#5]
# Reduce code duplication

There is currently a lot of code duplication between `Tree` and `TreeDeque` implementations.
An idea would be to have something akin to :
```rust
enum Tree<T, Container = Vec<Tree<T, Container>>> {
    Leaf(T),
    Branch(Container)
}

type TreeDeque<T> = Tree<T, VecDeque<TreeDeque<T>>>
```

**Comment by ThibSrb** [comment]
# Conclusion

**TLDR** : After some experimentation on the topic, i've decided not to go further in that direction and to keep the implementation as it is currently, even if it means to accept some code duplication.

## Problems

The envisaged idea is possible but not as presented in the issue.
In fact, this code does not compile :
```rust
enum Tree<T, Container = Vec<Tree<T, Container>>> {
    Leaf(T),
    Branch(Container)
}
```
> generic parameters with a default cannot use forward declared identifiers
> defaulted generic parameters cannot be forward declared

and the type aliases idea isn't possible either :
```rust
type TreeDeque<T> = Tree<T, VecDeque<TreeDeque<T>>>
```
> cycle detected when expanding type alias `TreeDeque`
> ...which immediately requires expanding type alias `TreeDeque` again
> type aliases cannot be recursive

## Solution

To make this work, it is necessary to use a wrapping structure such as :

```rust
struct VecTree<T>(Tree<T, Vec<VecTree<T>>>);
```
An example implementation of VecTree is available on the branch corresponding to this issue.

## Decision

The solution adds nesting to the generated structure (each tree being nested in a VecTree) and adds useless complexity to the project, so I decided not to go further in that direction and to and stay with the current implementation.
